0016-pre-shared-key/0016-pre-shared-key.md

```diff
@@ -0,0 +1,77 @@
+
+An advantage of PSK is that the pre-shared secret key only needs to be shared once. So long as both parties possess this key and are listening for transfers, they can send payments between one another.
+
+A disadvantage of PSK is that it is repudiable. Knowledge of a payment's fulfillment cannot be used as cryptographic proof that the receiver completed the payment, because both sender and receiver know how to generate the fulfillment with their pre-shared secret key. This causes no issues unless you have some other framework on top of ILP that relies on this cryptographic proof.
+
+## Flow
+
+0. The pre-shared secret key is shared out of band.
+0. The sender creates an inner (private) PSK headers, with information on the expiry and any additional data enclosed within.
+0. The sender creates a random token value, and calculates an HMAC of the token using the pre-shared secret key as the key. This HMAC will be referred to as the payment key.
+0. The sender encrypts the inner headers and their data with the payment key using AES-256-CTR.
+0. The sender creates a PSK packet containing the outer (public) headers, and attaches the token and HMAC function generated the payment key with as headers. The encrypted inner headers are attached as binary data.
+0. The sender quotes to the receiver and sends a transfer with the outer headers in the data field of the ILP packet. The fulfillment of this payment is the HMAC of the ILP packet using the payment key as the key.
+0. The receiver gets notified by their ledger of an incoming prepared transfer with an ILP packet.
+0. The receiver generates the payment key from the outer headers contained in the data field of the ILP packet, and uses it to decrypt the inner headers and their data.
+0. The receiver verifies that the amount in the incoming transfer matches the amount in the ILP packet. The receiver may also call a backend to make sure the incoming funds are expected.
+0. The receiver fulfills the incoming transfer with the HMAC of the ILP packet, using the payment key as the key.
+
+## Data Format
+
+The PSK details are divided into two sets of headers, like those found in [HTTP Requests](https://tools.ietf.org/html/rfc7230#section-3.2). Custom binary data is appended to the inner headers, after a blank line. This object is then encrypted and appended to the outer headers as binary data, after a blank line. Both inner and outer headers are parsed in the exact same way as HTTP headers.
+
+The outer headers come after a single status line, which must be `PSK/1.0`. The outer headers are visible to all parties transmitting the packet, so they SHOULD contain only the bare minimum required for the intended parties to decrypt and validate the inner headers.
```

Comment:
I'm gonna be optimistic and say that the prose is correct and the example is wrong, because that's the version I like. ;)